Add native RTF support to wxTextCtrl (macOS)

Blake-Madden opened the pull request:
This adds:

- `GetRTFValue()` and `SetRTFValue()` functions to get and set raw RTF
- Changes `wxTEXT_TYPE_ANY` into a `wxTextCtrlFileType` enum that includes a new `wxTEXT_TYPE_RTF` option
- `LoadFile()` and `SaveFile()` now use their `fileType` parameter to save as either text or RTF (if supported)
- Demonstrates these new feature in the "text" sample
- Simple unit test

P.S.
An MSW implementation for the RICHEDIT2 control will be forthcoming in another PR if this is accepted.

Blake-Madden committed:
Add Get and Set RTF functions to wxTextCtrl under macOS

Blake-Madden committed:
Add RTF option to save and load methods for wxTextCtrl
Supported under macOS currently; fails elsewhere

Blake-Madden committed:
Add demonstration (and testing) of RTF to text sample

Blake-Madden committed:
Make GetRTFValue() and SetRTFValue() virtual
Default is to return empty string or do nothing (with asserts) if not supported by the control.

Blake-Madden committed:
Fix compile issue and log info about RTF file when saving in text sample

Blake-Madden committed:
Fix asserts

Blake-Madden committed:
Fix misspelling

Blake-Madden committed:
Fix whitespace

Blake-Madden committed:
Fix whitespace

Blake-Madden committed:
Add unit test

Blake-Madden committed:
Fix unit test

Blake-Madden committed:
Make wxTEXT_TYPE_ANY use the file extension when determining file format
Before it was also text, but now will take into account the user providing a filename with an RTF extension (on macOS only for now).

Blake-Madden commented:
I think the AppVeyor test failure was just a sporadic issue.

vadz commented on include/wx/osx/cocoa/private/textimpl.h:
```diff
@@ -60,6 +60,15 @@ public :
+    {
+        wxFAIL_MSG("GetRTFValue() should only be used with multiline controls.");
+        return wxEmptyString;
+    };
```
Minor, but this will result in warnings with gcc `-pedantic`:

```suggestion
    }
```

Blake-Madden replied:
Got it.

vadz commented on include/wx/osx/cocoa/private/textimpl.h:
```diff
@@ -60,6 +60,15 @@ public :
+    virtual void SetRTFValue(const wxString& WXUNUSED(str)) override
+    {
+        wxFAIL_MSG("SetRTFValue() should only be used with multiline controls.");
+    };
```
```suggestion
    }
```

Blake-Madden replied:
Got it.

vadz commented on include/wx/textctrl.h:
```diff
@@ -103,7 +103,12 @@ const wxTextCoord wxInvalidTextCoord    = -2;
 // ----------------------------------------------------------------------------
 
-#define wxTEXT_TYPE_ANY     0
+enum wxTextCtrlFileType
```
I think we ought to use scoped enum for this one.

Blake-Madden replied:
I was wanting to do that originally, but that would break the API for `LoadFile(const wxString&, int)` and `SaveFile(const wxString&, int)`; the `int` would need to become a `wxTextCtrlFileType`. I could do some static_cast'ing inside these functions, but that defeats the purpose of using a scoped enum. Thoughts?

vadz replied:
It's difficult to know if anybody really passes an `int` to these functions, knowing that there was just a single possible value for it so far, but I guess you're right and somebody must be doing it or, worse, passing just `0` here, so I take this back, let's keep your original approach, sorry.

vadz commented on include/wx/textctrl.h:
```diff
@@ -680,6 +685,17 @@ class WXDLLIMPEXP_CORE wxTextAreaBase
+    // overridden implementations are required to actually handle RTF
+    virtual wxString GetRTFValue() const
+    {
+        wxFAIL_MSG("GetRTFValue() should only be used with rich, multiline controls.");
```
Is this message really correct? I think it should say "not implemented for the current platform".

I'd also move this (and the function below) into `src/common/textcmn.cpp`, there doesn't seem to be any good reason to define them inline.

Blake-Madden replied:
Got it.

vadz commented on include/wx/textctrl.h:
```diff
@@ -680,6 +685,17 @@ class WXDLLIMPEXP_CORE wxTextAreaBase
+        wxFAIL_MSG("GetRTFValue() should only be used with rich, multiline controls.");
+        return wxEmptyString;
+    }
+    virtual void SetRTFValue(const wxString& WXUNUSED(val))
```
I think this should return `bool` as it looks like there could be errors here — e.g. what if the RTF is syntactically invalid?

Blake-Madden replied:
The underlying function returns void:

https://developer.apple.com/documentation/appkit/nstext/1527249-replacecharactersinrange

and I've never seen an error state raised when I pass in malformed RTF. It just silently does what it can with the content it seems.  The RICHEDIT2 control behaves the same way.

vadz commented on interface/wx/textctrl.h:
```diff
@@ -1623,6 +1641,37 @@ class wxTextCtrl : public wxControl,
+
+        This offers more granular control of content formatting, as well as a
+        significant performance benefit with larger content. This also provides the
+        ability read an RTF file and move it directly into the control.
```
```suggestion
        ability to read an RTF file and move it directly into the control.
```

Blake-Madden replied:
Got it.

vadz commented on samples/text/text.cpp:
```diff
@@ -1668,6 +1698,35 @@ void MyFrame::OnFileLoad(wxCommandEvent& WXUNUSED(event))
+        // verify that the file length is correct
+        wxFile file("dummy.rtf");
+        wxLogStatus(this,
+            "Successfully saved file (text len = %lu, file size = %ld)",
+            (unsigned long)m_panel->m_textrich->GetValue().length(),
+            (long)file.Length());
```
We should be able to avoid casts here:
```suggestion
            "Successfully saved file (text len = %zu, file size = %lld)",
            m_panel->m_textrich->GetValue().length(),
            file.Length());
```

Blake-Madden replied:
Got it

vadz reviewed (commented):
Thanks, this looks good, but I believe we might want to add some way of checking whether RTF support is available, either at compile- or run-time, i.e. it could be as simple as defining `wxHAS_RTF_IN_TEXTCTRL` or adding `static bool wxTextCtrl::IsRTFSupported()`, what do you think?

Blake-Madden commented:
> Thanks, this looks good, but I believe we might want to add some way of checking whether RTF support is available, either at compile- or run-time, i.e. it could be as simple as defining `wxHAS_RTF_IN_TEXTCTRL` or adding `static bool wxTextCtrl::IsRTFSupported()`, what do you think?

I'm liking `static bool wxTextCtrl::IsRTFSupported()` myself; it's more modern than another `#define` :)

Blake-Madden committed:
Remove stray semicolons

Blake-Madden committed:
Remove casts

Blake-Madden committed:
Fix typo

Blake-Madden committed:
Improve error message, move to implementation file

Blake-Madden committed:
Add IsRTFSupported()

Blake-Madden commented:
OK, I think everything is ready now. Thanks.

PBfordev commented on samples/text/text.cpp:
```diff
@@ -283,6 +283,8 @@ class MyFrame: public wxFrame
 #endif // wxUSE_LOG
     void OnFileSave(wxCommandEvent& event);
     void OnFileLoad(wxCommandEvent& event);
+    void OnFileSaveRtf(wxCommandEvent& event);
```
Why is in these two methods RTF spelled "Rtf", unlike everywhere else?

Blake-Madden replied:
Fixed, thanks.

PBfordev commented on samples/text/text.cpp:
```diff
@@ -1668,6 +1698,35 @@ void MyFrame::OnFileLoad(wxCommandEvent& WXUNUSED(event))
+        wxFile file("dummy.rtf");
+        wxLogStatus(this,
+            "Successfully saved file (text len = %lu, file size = %ld)",
+            (unsigned long)m_panel->m_textrich->GetValue().length(),
```
The casts have not been resolved here, only in OnFileSave()?

Blake-Madden replied:
Fixed, thanks.

Blake-Madden committed:
Remove casts, make function names consistent

vadz committed:
Tweaks to RTF support

Define wxHAS_TEXTCTRL_RTF instead of testing for wxOSX.

Rename wxTEXT_TYPE_TEXT to wxTEXT_TYPE_PLAIN and make its value
different from wxTEXT_TYPE_ANY.

Use "switch" to ensure that we handle all wxTextCtrlFileType values.

Minor style/formatting fixes.

vadz commented:
Thanks for the updates!

I've pushed some minor changes, repeating the commit message (as it will get squashed here):

- Define `wxHAS_TEXTCTRL_RTF` instead of testing for wxOSX.
- Rename `wxTEXT_TYPE_TEXT` to `wxTEXT_TYPE_PLAIN` and make its value different from `wxTEXT_TYPE_ANY`.
- Use `switch` to ensure that we handle all `wxTextCtrlFileType` values.
- Minor style/formatting fixes.

Please let me know if you have any objections to/comments on any of them. Otherwise I think we should be able to merge this.

Thanks again for your work on this!

vadz committed:
Document wxHAS_TEXTCTRL_RTF

vadz committed:
Merge branch 'master' into rtf-textctrl

vadz commented:
FYI, I've also merged master into this branch and resolved the conflict in the change log to let the CI builds run.

Blake-Madden commented:
> Thanks for the updates!
> 
> I've pushed some minor changes, repeating the commit message (as it will get squashed here):
> 
>     * Define `wxHAS_TEXTCTRL_RTF` instead of testing for wxOSX.
> 
>     * Rename `wxTEXT_TYPE_TEXT` to `wxTEXT_TYPE_PLAIN` and make its value different from `wxTEXT_TYPE_ANY`.
> 
>     * Use `switch` to ensure that we handle all `wxTextCtrlFileType` values.
> 
>     * Minor style/formatting fixes.
> 
> 
> Please let me know if you have any objections to/comments on any of them. Otherwise I think we should be able to merge this.
> 
> Thanks again for your work on this!

These all sound great, no objections here. Thanks!
